src/streams/register_nack.rs

```diff
@@ -0,0 +1,560 @@
+use std::ops::Range;
+
+use crate::rtp_::{Nack, NackEntry, ReportList, SeqNo};
+
+/// How many sequence numbers we keep track of
+const MAX_DROPOUT: u64 = 3000;
+
+/// Number of out of order packets we keep track of for reports
+/// A suggested maximum value min is MAX_DROPOUT / 2
+const MAX_MISORDER: u64 = 100;
+
+const U16_MAX: u64 = u16::MAX as u64 + 1_u64;
+
+/// The max number of NACKs we perform for a single packet
+const MAX_NACKS: u8 = 5;
+
+#[derive(Debug)]
+pub struct NackRegister {
+    /// Status of packets indexed by wrapping SeqNo.
+    packets: Vec<PacketStatus>,
+
+    /// Range of seq numbers considered NACK reporting.
+    active: Option<Range<SeqNo>>,
+}
+
+#[derive(Debug, Default, Clone, Copy)]
+struct PacketStatus {
+    received: bool,
+    nack_count: u8,
+}
+
+impl PacketStatus {
+    fn needs_nack(&self) -> bool {
+        !self.received && self.nack_count < MAX_NACKS
+    }
+
+    fn mark_received(&mut self) -> bool {
+        let new = !self.received;
+        self.received = true;
+        new
+    }
+
+    fn reset(&mut self) {
+        self.received = false;
+        self.nack_count = 0;
+    }
+}
+
+impl NackRegister {
+    pub fn new() -> Self {
+        NackRegister {
+            packets: vec![PacketStatus::default(); MAX_DROPOUT as usize],
+            active: None,
+        }
+    }
+
+    pub fn update(&mut self, seq: SeqNo) -> bool {
+        let Some(active) = self.active.clone() else {
```

Comment:
`Range` not being `Copy` when `Idx` is is one of those weird quirks of Rust. If we aren't heavily relying on the Range functionality we could create our own type that could be `Copy` instead

Reply:
Not needing to clone here and also not needing to do the `as_ref` business elsewhere